src/OpenTelemetry.Api/Baggage.cs

```diff
@@ -0,0 +1,305 @@
+            }
+
+            return this.baggage != null && this.baggage.TryGetValue(name, out string value)
+                ? value
+                : null;
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="Baggage"/> which contains the new key/value pair.
+        /// </summary>
+        /// <param name="name">Baggage item name.</param>
+        /// <param name="value">Baggage item value.</param>
+        /// <returns>New <see cref="Baggage"/> containing the key/value pair.</returns>
+        public Baggage SetBaggage(string name, string value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                return this.RemoveBaggage(name);
+            }
+
+            return Current = new Baggage(
+                new Dictionary<string, string>(this.baggage ?? EmptyBaggage, StringComparer.OrdinalIgnoreCase)
+                {
+                    [name] = value,
+                });
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="Baggage"/> which contains the new key/value pair.
+        /// </summary>
+        /// <param name="baggageItems">Baggage key/value pairs.</param>
+        /// <returns>New <see cref="Baggage"/> containing the key/value pair.</returns>
+        public Baggage SetBaggage(params KeyValuePair<string, string>[] baggageItems)
+            => this.SetBaggage((IEnumerable<KeyValuePair<string, string>>)baggageItems);
+
+        /// <summary>
+        /// Returns a new <see cref="Baggage"/> which contains the new key/value pair.
+        /// </summary>
+        /// <param name="baggageItems">Baggage key/value pairs.</param>
+        /// <returns>New <see cref="Baggage"/> containing the key/value pair.</returns>
+        public Baggage SetBaggage(IEnumerable<KeyValuePair<string, string>> baggageItems)
+        {
+            if ((baggageItems?.Count() ?? 0) <= 0)
+            {
+                return this;
+            }
+
+            var newBaggage = new Dictionary<string, string>(this.baggage ?? EmptyBaggage, StringComparer.OrdinalIgnoreCase);
+
+            foreach (var item in baggageItems)
+            {
+                if (string.IsNullOrEmpty(item.Value))
+                {
+                    newBaggage.Remove(item.Key);
+                }
+                else
+                {
+                    newBaggage[item.Key] = item.Value;
+                }
+            }
+
+            return Current = new Baggage(newBaggage);
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="Baggage"/> with the key/value pair removed.
+        /// </summary>
+        /// <param name="name">Baggage item name.</param>
+        /// <returns>New <see cref="Baggage"/> containing the key/value pair.</returns>
+        public Baggage RemoveBaggage(string name)
+        {
+            var baggage = new Dictionary<string, string>(this.baggage ?? EmptyBaggage, StringComparer.OrdinalIgnoreCase);
```

Comment:
minor: if the key does not exist, we might be able to return the same baggage?

Reply:
I guess it is not important at this stage, and we probably will replace dictionary with a better data structure before GA anyways.

Reply:
OK I'll leave as is. I did consider that case. I figured if user is calling RemoveBaggage, they probably know it exists. So I didn't want to waste perf doing a hash + lookup where in most cases it would be wasteful? My thinking could be off tho.